### PR TITLE
HV-1767 Reduce the overhead of ConstraintHelper initialization in the predefined scope case

### PR DESCRIPTION
**Pull request by gsmet**
https://hibernate.atlassian.net/browse/HV-1767

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/metadata/core/ConstraintHelper.java [hunk 1]
```diff
@@ -1092,16 +1097,28 @@ public void clear() {
...
 	}
 
-	private static boolean isJodaTimeInClasspath() {
-		return isClassPresent( JODA_TIME_CLASS_NAME );
+	private boolean isJodaTimeInClasspath() {
+		if ( jodaTimeInClassPath != null ) {
+			return jodaTimeInClassPath.booleanValue();
+		}
+		jodaTimeInClassPath = isClassPresent( JODA_TIME_CLASS_NAME );
+		return jodaTimeInClassPath.booleanValue();
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/validator/internal/metadata/core/ConstraintHelper.java:1105]: Some nitpicking (for a change!); I usually see this pattern written that way:

```
		if ( jodaTimeInClassPath == null ) {
			jodaTimeInClassPath = isClassPresent( JODA_TIME_CLASS_NAME );
		}
		return jodaTimeInClassPath.booleanValue(); 
```

File: engine/src/main/java/org/hibernate/validator/internal/metadata/core/BuiltinConstraint.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+				Set<BuiltinConstraint> constraints = new HashSet<>();
+				constraints.add( constraint );
+				constraints.addAll( constraint.composingConstraints );
+				CONSTRAINT_MAPPING.put( constraint.constraint, constraints );
+			}
+		}
+	}
+
+	private String constraint;
+	private List<BuiltinConstraint> composingConstraints;
```
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/validator/internal/metadata/core/BuiltinConstraint.java:110]: `requiredConstraints` or `constraintDependencies` or something like that? Feel free to ignore, but I had a hard time wrapping my head around this, so a more precise name might help.

Comment (gsmet) [reply engine/src/main/java/org/hibernate/validator/internal/metadata/core/BuiltinConstraint.java:110]: This is the technical term in HV but yeah I also thought about changing it to something along the lines of what you suggest so let's do it.

**Comment by gsmet** [comment]
@yrodiere I addressed your comments.

**Pull request merged**